httpbp/client_middlewares.go

```diff
@@ -0,0 +1,240 @@
+package httpbp
+
+import (
+	"errors"
+	"io"
+	"net/http"
+	"sync"
+	"sync/atomic"
+
+	retry "github.com/avast/retry-go"
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/reddit/baseplate.go/breakerbp"
+	"github.com/reddit/baseplate.go/retrybp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+const (
+	// DefaultMaxErrorReadAhead defines the maximum bytes to be read from a
+	// failed HTTP response to be attached as additional information in a
+	// ClientError response.
+	DefaultMaxErrorReadAhead = 1024
+)
+
+// ClientMiddleware is used to build HTTP client middleware by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(next http.RoundTripper) http.RoundTripper
+
+// roundTripperFunc adapts closures and functions to implement http.RoundTripper.
+type roundTripperFunc func(req *http.Request) (*http.Response, error)
+
+func (f roundTripperFunc) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+// NewClient returns a standard HTTP client wrapped with the default middleware
+// plus any additional client middleware passed into this function. Default
+// middlewares are: MonitorClient and Retries. ClientErrorWrapper is included
+// as transitive middleware through Retries.
+func NewClient(config ClientConfig, middleware ...ClientMiddleware) (*http.Client, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+
+	// set max connections per host if set
+	var transport http.Transport
+	if config.MaxConnections > 0 {
+		transport.MaxConnsPerHost = config.MaxConnections
+	}
+
+	// apply default if not set
+	if config.MaxErrorReadAhead == 0 {
+		config.MaxErrorReadAhead = DefaultMaxErrorReadAhead
+	}
+
+	defaults := []ClientMiddleware{
+		MonitorClient(config.Slug),
+		ClientErrorWrapper(config.MaxErrorReadAhead),
+	}
+	// apply retries only if retry options are set but replace
+	// ClientErrorWrapper since Retries already wraps ClientErrorWrapper
+	if len(config.RetryOptions) > 0 {
+		defaults[1] = Retries(
+			config.MaxErrorReadAhead,
+			config.RetryOptions...,
+		)
+	}
```

Comment:
We should not be adding `ClientErrorWrapper` to default middlewares, and we should always add `Retries` to default middlewares, even if `config.RetryOptions` is not set.

Can you follow [`thriftbp`'s example](https://pkg.go.dev/github.com/reddit/baseplate.go/thriftbp#BaseplateDefaultClientMiddlewares)?